File: car.cpp
#include "car.h"
#include "wheel.h"
#include <QPainter>
#include <QtMath> // For M_PI and atan2

Car::Car(Wheel* backWheel, Wheel* frontWheel)
    : m_backWheel(backWheel), m_frontWheel(frontWheel) {
    loadImage();
}

void Car::loadImage() {
    // Load the image from Qt's resource system.
    // Make sure you have a .qrc file in your project!
    m_image.load(":/shino/car.png");
    if (m_image.isNull()) {
        qWarning("Failed to load car.png from resources!");
    }
}

void Car::update(bool accelerating, bool braking) {
    if (!m_backWheel || !m_frontWheel) return;

    // Apply torque by directly changing wheel vertical velocities
    if (accelerating) {
        m_backWheel->m_vy -= TORQUE_FORCE;
        m_frontWheel->m_vy += TORQUE_FORCE;
    }
    if (braking) {
        m_backWheel->m_vy += TORQUE_FORCE;
        m_frontWheel->m_vy -= TORQUE_FORCE;
    }

    // Calculate angle based on the physical positions of the wheels
    m_angle = qAtan2(m_frontWheel->y - m_backWheel->y, m_frontWheel->x - m_backWheel->x);

    // Check if the car is flipped (angle > 90 degrees)
    if (qAbs(m_angle) > M_PI / 2.0) {
        m_isFlipped = true;
    }

    // Calculate draw scale based on wheel size
    double rAvg = (m_backWheel->getRadius() + m_frontWheel->getRadius()) * 0.5;
    m_drawScale = rAvg / IMG_WHEEL_RADIUS;
}

bool Car::isFlipped() const {
    return m_isFlipped;
}

void Car::draw(QPainter& p, int camX, int camY) {
    if (m_image.isNull() || !m_backWheel) return;

    p.save(); // Save painter state

    // Get screen coordinates of the back wheel
    const double screenBackX = m_backWheel->x - camX;
    const double screenBackY = m_backWheel->y + camY;

    // Apply transformations (Translate, Rotate, Scale)
    p.translate(screenBackX, screenBackY);
    p.rotate(qRadiansToDegrees(m_angle)); // qAtan2 is in radians, rotate is in degrees
    p.scale(m_drawScale, m_drawScale);
    p.translate(-IMG_BACK_CX, -IMG_BACK_CY + m_fineDY);

    // Draw the car image
    p.drawImage(0, 0, m_image);

    p.restore(); // Restore painter state
}

File: shino/Car.java
package shino;

import javax.imageio.ImageIO;
import java.awt.Graphics2D;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.io.InputStream;

public class Car {
    private final Wheel backWheel, frontWheel;
    private BufferedImage image;

    public static final double IMG_BACK_CX      = 36;   // px
    public static final double IMG_BACK_CY      = 112;  // px
    public static final double IMG_FRONT_CX     = 114;  // px
    public static final double IMG_FRONT_CY     = 112;  // px
    public static final double IMG_WHEEL_BASE   = IMG_FRONT_CX - IMG_BACK_CX; // px
    public static final double IMG_WHEEL_RADIUS = 12;   // px

    private final double TORQUE_FORCE = 0.05;

    private double angle;
    private double drawScale = 1.0;
    private double fineDX = 0, fineDY = 18;

    public void nudgeInCarFrame(double dx, double dy){
        fineDX += dx; fineDY += dy;
    }

    public Car(Wheel backWheel, Wheel frontWheel) {
        this.backWheel = backWheel;
        this.frontWheel = frontWheel;
        loadImage();
    }

    private void loadImage() {
        try (InputStream in = Car.class.getResourceAsStream("/shino/car.png")) {
            if (in == null) throw new IllegalStateException("car.png not found on classpath");
            image = ImageIO.read(in);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void update(boolean accelerating, boolean braking) {
        if (accelerating) {
            backWheel.vy -= TORQUE_FORCE;
            frontWheel.vy += TORQUE_FORCE;
        }
        if (braking) {
            backWheel.vy += TORQUE_FORCE;
            frontWheel.vy -= TORQUE_FORCE;
        }

        angle = Math.atan2(frontWheel.y - backWheel.y, frontWheel.x - backWheel.x);

        double rAvg = (backWheel.getRadius() + frontWheel.getRadius()) * 0.5;
        drawScale = rAvg / IMG_WHEEL_RADIUS;
    }

    public int getVisualWheelRadius() {
        return (int) Math.round(IMG_WHEEL_RADIUS * drawScale);
    }
    
    public void draw(Graphics2D g2d, int camX, int camY) {
        if (image == null) return;

        int screenBackX = (int)Math.round(backWheel.x - camX);
        int screenBackY = (int)Math.round(backWheel.y + camY);

        AffineTransform at = new AffineTransform();
        at.translate(screenBackX, screenBackY);
        at.rotate(angle);
        at.scale(drawScale, drawScale);
        at.translate(-IMG_BACK_CX + fineDX, -IMG_BACK_CY + fineDY);

        g2d.drawImage(image, at, null);
    }
}